### PR TITLE
[flutter_tools] Support Integration Tests

### PR DESCRIPTION
**Pull request by jiahaog**
Implements https://flutter.dev/go/tool-integration-test-support

### Usage

Examples below are not exhaustive. Flags compatible with `flutter test` should work as well.

```sh
# Run an Integration Test.
flutter test integration_test/some_test.dart

# Run an Integration Test on the specified device.
flutter -d my-device-id test integration_test/some_test.dart

# Run all tests in `integration_test/` as Integration Tests.
flutter test integration_test
```

Minor changes:

- Translate received test file paths to absolute paths before passing them to the `package:test` runner
- Some tweaking of the `FakeDevice` to support mocking a `LaunchResult`

### Upcoming PR

- Support web
- Device lab tests (#79258) 

Related issue: #66264

### CODE REVIEW

File: packages/flutter_tools/lib/src/test/test_device.dart
Comment (jiahaog) [inline packages/flutter_tools/lib/src/test/test_device.dart:21]: Unfortunately the abstraction here isn't very ideal.

For flutter_test, we can accept the precompiled dill, or if we move the compiler into `FlutterTesterTestDevice`, we can accept Dart source files.

For a raw `Device`, it accepts a source file, or a precompiled APK/IPA.

I can't quite think of a way to abstract this better given the awkward compilation of sources that happen in the `FlutterTestPlatform` for the `flutter_tester`. let me know if there's any better ideas.

**Comment by jiahaog** [comment]
@jonahwilliams Let me know if you have any early feedback for this - Most of the problems I've been having here have been resolved. (Still working on adding the tests)

File: packages/flutter_tools/lib/src/commands/test.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/test.dart:225]: nit: indentation

File: packages/flutter_tools/lib/src/commands/test.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/test.dart:247]: same here

File: packages/flutter_tools/lib/src/commands/test.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/test.dart:26]: Consider extending the description of the test command to describe the difference in behavior between test and integration test commands

File: packages/flutter_tools/lib/src/test/integration_test_device.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/test/integration_test_device.dart:70]: There may be some timing issues here, since the kIntegrationTestMethod will not be registered until after main finishes, but this may be after the app/isolate has started. There is a waitForExtension method in the resident_runner library that handles this cases, though it doesn't identify the correct Isolate:

https://github.com/flutter/flutter/blob/dea1c46a3c046766089a5d2e85bf6e169411b82e/packages/flutter_tools/lib/src/resident_devtools_handler.dart#L186

File: packages/flutter_tools/lib/src/test/test_device.dart
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/test/test_device.dart:21]: I think its OK, provided you document the convention.

File: packages/flutter_tools/lib/src/test/integration_test_device.dart
Comment (jiahaog) [reply packages/flutter_tools/lib/src/test/integration_test_device.dart:70]: Would there be issues with the implementation of `_findExtensionIsolate` method defined below? It looks like it's implementing something similar to the `waitForExtension` method (both subscribing to registered extension events as well as querying for current extensions, which should solve the timing issues here

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/test/integration_test_device.dart:70]: Yeah that makes sense - I would prefer trying to reuse the existing helper if possible.

File: packages/flutter_tools/lib/src/test/test_device.dart
Comment (jiahaog) [reply packages/flutter_tools/lib/src/test/test_device.dart:21]: Updated the doc comment for this

File: packages/flutter_tools/lib/src/commands/test.dart
Comment (jiahaog) [reply packages/flutter_tools/lib/src/commands/test.dart:63]: Updated the doc comment for this

File: packages/flutter_tools/lib/src/test/integration_test_device.dart
Comment (jiahaog) [reply packages/flutter_tools/lib/src/test/integration_test_device.dart:47]: Yes, I think we could make `flutter_platform.dart` keep track of some sort of mutable state where rebuilding is skipped for subsequent workflows. For the first run we basically do what is here in this PR, and for subsequent runs we would invoke portions of the attach workflow, to attach and then hot restart the application. I can't be sure if it's sufficient without a prototype, but it seems like a `reloadExistingApp` boolean parameter here would be sufficient. It's also fine if we want to make breaking changes to the interface here because it's all internal anyway.

File: packages/flutter_tools/lib/src/test/integration_test_device.dart
Comment (jiahaog) [reply packages/flutter_tools/lib/src/test/integration_test_device.dart]: Refactoring to make the existing helper reusable here was done in #77220, and just updated this PR to use the helper here.

**Comment by jiahaog** [comment]
Sorry for the delay, I was caught up with some other things, and just found some time to clean this PR up and add more tests. It is finally ready for review 🙂 

(I also have a follow up PR #79258 to add device lab tests)

File: packages/flutter_tools/lib/src/test/integration_test_device.dart
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/test/integration_test_device.dart:47]: Sounds reasonable to me

**Review by jonahwilliams — COMMENTED** [review]
Overall this LGTM.

How comfortable do you feel with these changes, do you think we should wait until the next release cut, or until the devicelab test is ready?

As long as the existing `flutter test` behavior` is unchanged it should be OK

**Comment by jiahaog** [comment]
I think it should be fine to land this without waiting for the next release cut. The only change to the normal `flutter test` flow is the change to use canonical paths which didn't work on windows when debugging, so I've just updated the PR to use absolute paths. Also just manually tested adding a breakpoint and expression evaluation of the widget test flow on macOS, Linux and Windows in VSCode, and it seems to work fine. Is there anything else I should test manually?

Also, I figured out how to test the device lab changes locally in #79258 (found an issue and updated this PR to disable port publication like how `drive` does it), and we should be good to land the device lab change right after this PR lands. Splitting it off into its own PR so that would be easier to revert, as I'm a bit unsure of whether `dev/prod_builders.json` is specified correctly there.

**Comment by DanTup** [comment]
FWIW, I have an [open PR](https://github.com/Dart-Code/Dart-Code/pull/3133) in Dart-Code to support this. I don't plan to merge it for the upcoming release (this week) as I wanted to re-test it all once this PR is complete/merged (it also needs some automated tests). I'll likely merge it just after my release though (and I can make a preview build of the extension that includes the changes for testing/using in the meantime).

File: packages/flutter_tools/lib/src/commands/test.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/test.dart:255]:
```suggestion
    _testFiles = argResults.rest.map<String>(globals.fs.path.absolute).toList();
```

**Review by jonahwilliams — APPROVED** [review]
only have a nit.

LGTM

**Comment by jiahaog** [comment]
@DanTup sure, no hurry. I'll merge this and then try and patch your change as well to test it out

**Pull request merged**
